### PR TITLE
opt: add SplitDisjunctionAddKey exploration rule for disjunctions

### PR DESCRIPTION
**Pull request by mgartner**
This commit adds a new exploration rule that can produce better query
plans for disjunctions (e.g. a = 1 OR b = 2). It is similar to
SplitDisjunction, but handles the special case when the selected
columns of a query do not contain a strict key.

Consider the query below, where a and b are not primary keys.

  SELECT a, b FROM t WHERE a = 1 OR b = 2

Before this commit, this disjunction would not be optimized into a UNION
of two SELECT expressions, because the SplitDisjunction exploration
rule only applies when the output columns contain a strict key.

Release note (performance improvement): Query optimization for
disjunctions was extended to support queries which do not select strict
keys.

### CODE REVIEW

**Review by mgartner — COMMENTED** [review]
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/47342)* status: :shipit: complete! 0 of 0 LGTMs obtained

___
*[pkg/sql/opt/xform/rules/select.opt, line 108 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/47342#-M4_33o8ANPad4pEC7Z1:-M4_33o8ANPad4pEC7Z2:bav6rgy) ([raw file](https://github.com/cockroachdb/cockroach/blob/4d6c98393d729a387bd3e1453ec41e2b400ead78/pkg/sql/opt/xform/rules/select.opt#L108)):*
> ```
> # guaranteed to be unique. This prevents incorrect de-duplication and guarantees
> # that the newly generated plan is equivalent to the original plan.
> [GenerateUnionSelectsAddKey, Explore]
> ```

I'm not in love with this rule name, but I thought something like `GenerateUnionSelectsWithKey` or `GenerateUnionSelectsWithoutKey` were both unclear whether they mean the original expression has a key or the new expression will have a key. Open to any suggestions...

**Review by RaduBerinde — COMMENTED** [review]
Cool!

> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/47342#-:-M4aADv_5LcxGFsxCFxY:b-89q77x)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @mgartner)

___
*[pkg/sql/opt/xform/rules/select.opt, line 108 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/47342#-M4_33o8ANPad4pEC7Z1:-M4a9BOrBdA1zf2Y0Ft3:bc96enc) ([raw file](https://github.com/cockroachdb/cockroach/blob/4d6c98393d729a387bd3e1453ec41e2b400ead78/pkg/sql/opt/xform/rules/select.opt#L108)):*
<details><summary><i>Previously, mgartner (Marcus Gartner) wrote…</i></summary><blockquote>

I'm not in love with this rule name, but I thought something like `GenerateUnionSelectsWithKey` or `GenerateUnionSelectsWithoutKey` were both unclear whether they mean the original expression has a key or the new expression will have a key. Open to any suggestions...
</blockquote></details>

Now that I think about it, `GenerateUnionSelects` is not a great name to begin with. It technically describes what it generates but there's no context about it. Maybe `SplitDisjunction` and `SplitDisjunctionAddKey`
___
*[pkg/sql/opt/xform/testdata/rules/select, line 2006 at r3](https://reviewable.io/reviews/cockroachdb/cockroach/47342#-M4aAvoA686ONKXvMS61:-M4aAvoA686ONKXvMS62:b-toh11j) ([raw file](https://github.com/cockroachdb/cockroach/blob/c04b7a4160d2c3b11c2cac026d92249de746c5ed/pkg/sql/opt/xform/testdata/rules/select#L2006)):*
> ```
> # index, the generated expression with a UNION has a higher cost and is not
> # part of the final expression.
> opt
> ```

this should still have an `expect=`

**Review by mgartner — COMMENTED** [review]
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/47342)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @RaduBerinde)

___
*[pkg/sql/opt/xform/rules/select.opt, line 108 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/47342#-M4_33o8ANPad4pEC7Z1:-M4afkp1Et4PpfwJxHG7:bkoke9u) ([raw file](https://github.com/cockroachdb/cockroach/blob/4d6c98393d729a387bd3e1453ec41e2b400ead78/pkg/sql/opt/xform/rules/select.opt#L108)):*
<details><summary><i>Previously, RaduBerinde wrote…</i></summary><blockquote>

Now that I think about it, `GenerateUnionSelects` is not a great name to begin with. It technically describes what it generates but there's no context about it. Maybe `SplitDisjunction` and `SplitDisjunctionAddKey`
</blockquote></details>

I like it. Done.
___
*[pkg/sql/opt/xform/testdata/rules/select, line 2006 at r3](https://reviewable.io/reviews/cockroachdb/cockroach/47342#-M4aAvoA686ONKXvMS61:-M4afp3c3a-nlR4n5pPe:b1bribj) ([raw file](https://github.com/cockroachdb/cockroach/blob/c04b7a4160d2c3b11c2cac026d92249de746c5ed/pkg/sql/opt/xform/testdata/rules/select#L2006)):*
<details><summary><i>Previously, RaduBerinde wrote…</i></summary><blockquote>

this should still have an `expect=`
</blockquote></details>

Good call. Unfortunately the addition of the check for interesting indexes now means this rule doesn't match. I need another way to assert that flags are preserved in the duplicate scan if I want to test that.

**Review by mgartner — COMMENTED** [review]
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/47342)* status: :shipit: complete! 0 of 0 LGTMs obtained

___
*[pkg/sql/opt/xform/testdata/rules/select, line 2006 at r3](https://reviewable.io/reviews/cockroachdb/cockroach/47342#-M4aAvoA686ONKXvMS61:-M4oZAxm-FHUF4Vbzmw-:b2ow55a) ([raw file](https://github.com/cockroachdb/cockroach/blob/c04b7a4160d2c3b11c2cac026d92249de746c5ed/pkg/sql/opt/xform/testdata/rules/select#L2006)):*
<details><summary><i>Previously, mgartner (Marcus Gartner) wrote…</i></summary><blockquote>

Good call. Unfortunately the addition of the check for interesting indexes now means this rule doesn't match. I need another way to assert that flags are preserved in the duplicate scan if I want to test that.
</blockquote></details>

I don't think there's anyway I can get the rule to fire when forcing a single index right now, based on its pattern. However, it may be possible in the future when I add support to this rule for inverted indexes. That should allow the rule to match when the columns on each side of the disjunction are equal, allowing the same index to be used for each side of the join. I'll try to add back tests like this then.

**Review by RaduBerinde — COMMENTED** [review]
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/47342)* status: :shipit: complete! 0 of 0 LGTMs obtained

___
*[pkg/sql/opt/xform/testdata/rules/select, line 2006 at r3](https://reviewable.io/reviews/cockroachdb/cockroach/47342#-M4aAvoA686ONKXvMS61:-M4odKd65_CFnAaoICQ8:bmrlu9l) ([raw file](https://github.com/cockroachdb/cockroach/blob/c04b7a4160d2c3b11c2cac026d92249de746c5ed/pkg/sql/opt/xform/testdata/rules/select#L2006)):*
<details><summary><i>Previously, mgartner (Marcus Gartner) wrote…</i></summary><blockquote>

I don't think there's anyway I can get the rule to fire when forcing a single index right now, based on its pattern. However, it may be possible in the future when I add support to this rule for inverted indexes. That should allow the rule to match when the columns on each side of the disjunction are equal, allowing the same index to be used for each side of the join. I'll try to add back tests like this then.
</blockquote></details>

There are other flags we can pass, e.g. `d@{NO_INDEX_JOIN}` prevents index join. We don't need the flag to be consequential to the plan, just need to see that both scans still have the flag.

**Review by rytaft — APPROVED** [review]
:lgtm: 

> Reviewed 1 of 2 files at r3, 4 of 4 files at r4.
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/47342#-:-M4pfTDr30gNlMJ4Xo5L:bg7igkv)* status: :shipit: complete! 1 of 0 LGTMs obtained (waiting on @mgartner)

___
*[pkg/sql/opt/exec/execbuilder/testdata/subquery, line 147 at r4](https://reviewable.io/reviews/cockroachdb/cockroach/47342#-M4pcDuj-wpXjqrqb0Tv:-M4pcDuj-wpXjqrqb0Tw:b6cs1u3) ([raw file](https://github.com/cockroachdb/cockroach/blob/415b842b701c17fbc97ec789001c08f7d6d4db25/pkg/sql/opt/exec/execbuilder/testdata/subquery#L147)):*
> ```
>  │                    render 0     col0             ·                                  ·
>  └── union            ·            ·                (col0, col3, col4, rowid[hidden])  ·
>       ├── norows      ·            ·                (col0, col3, col4, rowid)          ·
> ```

This plan could be further simplified by removing the UNION since one of the two inputs has no rows. Perhaps this could be another rule (in another PR, of course)?
___
*[pkg/sql/opt/xform/rules/select.opt, line 46 at r4](https://reviewable.io/reviews/cockroachdb/cockroach/47342#-M4pdjyaAwDYGG12g9rg:-M4pdjyb8E_UQP9Dd6SQ:bo8pfji) ([raw file](https://github.com/cockroachdb/cockroach/blob/415b842b701c17fbc97ec789001c08f7d6d4db25/pkg/sql/opt/xform/rules/select.opt#L46)):*
> ```
> #
> # Also note that this rule only matches Selects that have strict keys. See
> # SplitDisjunctionAddKey which handles this special case.
> ```

[nit] handles Selects that do not have strict keys.

**Review by rytaft — COMMENTED** [review]
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/47342)* status: :shipit: complete! 1 of 0 LGTMs obtained (waiting on @mgartner and @rytaft)

___
*[pkg/sql/opt/xform/testdata/rules/select, line 2006 at r3](https://reviewable.io/reviews/cockroachdb/cockroach/47342#-M4aAvoA686ONKXvMS61:-M4pgApn6BY-WC2YndMt:bnprr97) ([raw file](https://github.com/cockroachdb/cockroach/blob/c04b7a4160d2c3b11c2cac026d92249de746c5ed/pkg/sql/opt/xform/testdata/rules/select#L2006)):*
<details><summary><i>Previously, RaduBerinde wrote…</i></summary><blockquote>

There are other flags we can pass, e.g. `d@{NO_INDEX_JOIN}` prevents index join. We don't need the flag to be consequential to the plan, just need to see that both scans still have the flag.
</blockquote></details>

Did you add this test case somewhere else?

**Review by mgartner — COMMENTED** [review]
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/47342)* status: :shipit: complete! 0 of 0 LGTMs obtained (and 1 stale) (waiting on @mgartner and @rytaft)

___
*[pkg/sql/opt/xform/testdata/rules/select, line 2006 at r3](https://reviewable.io/reviews/cockroachdb/cockroach/47342#-M4aAvoA686ONKXvMS61:-M4pga8Q1dK6WA22rSxE:b-qdxgdm) ([raw file](https://github.com/cockroachdb/cockroach/blob/c04b7a4160d2c3b11c2cac026d92249de746c5ed/pkg/sql/opt/xform/testdata/rules/select#L2006)):*
<details><summary><i>Previously, rytaft (Rebecca Taft) wrote…</i></summary><blockquote>

Did you add this test case somewhere else?
</blockquote></details>

Just added those tests back in. Thanks for the NO_INDEX_JOIN flag tip @raduberinde, that was key.

**Review by rytaft — APPROVED** [review]
> Reviewed 1 of 1 files at r5.
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/47342)* status: :shipit: complete! 0 of 0 LGTMs obtained (and 1 stale) (waiting on @mgartner)

**Review by mgartner — COMMENTED** [review]
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/47342)* status: :shipit: complete! 0 of 0 LGTMs obtained (and 1 stale) (waiting on @rytaft)

___
*[pkg/sql/opt/exec/execbuilder/testdata/subquery, line 147 at r4](https://reviewable.io/reviews/cockroachdb/cockroach/47342#-M4pcDuj-wpXjqrqb0Tv:-M4phYpl8zaPDC0uj8GS:b-jwhtsh) ([raw file](https://github.com/cockroachdb/cockroach/blob/415b842b701c17fbc97ec789001c08f7d6d4db25/pkg/sql/opt/exec/execbuilder/testdata/subquery#L147)):*
<details><summary><i>Previously, rytaft (Rebecca Taft) wrote…</i></summary><blockquote>

This plan could be further simplified by removing the UNION since one of the two inputs has no rows. Perhaps this could be another rule (in another PR, of course)?
</blockquote></details>

Good catch. Would it make sense to make a normalization rule for this? And would it be best to look at all disjunctions and remove any that are guaranteed to return no rows? Or should this operate on Unions only?
___
*[pkg/sql/opt/xform/rules/select.opt, line 46 at r4](https://reviewable.io/reviews/cockroachdb/cockroach/47342#-M4pdjyaAwDYGG12g9rg:-M4ph3TO1rzxQutdM7vx:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/415b842b701c17fbc97ec789001c08f7d6d4db25/pkg/sql/opt/xform/rules/select.opt#L46)):*
<details><summary><i>Previously, rytaft (Rebecca Taft) wrote…</i></summary><blockquote>

[nit] handles Selects that do not have strict keys.
</blockquote></details>

Done.

**Review by rytaft — APPROVED** [review]
> Reviewed 1 of 1 files at r6.
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/47342)* status: :shipit: complete! 0 of 0 LGTMs obtained (and 1 stale) (waiting on @mgartner)

___
*[pkg/sql/opt/exec/execbuilder/testdata/subquery, line 147 at r4](https://reviewable.io/reviews/cockroachdb/cockroach/47342#-M4pcDuj-wpXjqrqb0Tv:-M4pivFX1pyKfx4b6Oah:b-skba2c) ([raw file](https://github.com/cockroachdb/cockroach/blob/415b842b701c17fbc97ec789001c08f7d6d4db25/pkg/sql/opt/exec/execbuilder/testdata/subquery#L147)):*
<details><summary><i>Previously, mgartner (Marcus Gartner) wrote…</i></summary><blockquote>

Good catch. Would it make sense to make a normalization rule for this? And would it be best to look at all disjunctions and remove any that are guaranteed to return no rows? Or should this operate on Unions only?
</blockquote></details>

Actually we already have this rule for `UnionAll` (see `set.opt` in the norm rules). We could add it for `Union`, but we'd still need to wrap it in a `Distinct` to eliminate duplicates. I think you're already making a separate rule to transform `Union` to `Distinct` + `UnionAll` anyway, so probably not worth it. (Also, I'm not sure the current normalization rule would even apply in this case since we don't currently apply normalization rules to the root of the tree in exploration rules).

So long story short... you can ignore this suggestion :)

**Review by RaduBerinde — APPROVED** [review]
:lgtm: Nice work!

> Reviewed 1 of 2 files at r3, 2 of 4 files at r4, 1 of 1 files at r5, 1 of 1 files at r6.
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/47342#-:-M4tU0360D65OadpSIR8:bcmw7ss)* status: :shipit: complete! 1 of 0 LGTMs obtained (and 1 stale) (waiting on @mgartner)
